### ISSUE TITLE
Misleading error message with precedence issue involving "as"

### ISSUE THREAD

**Issue by joshtriplett** [rust-lang/rust#36367]
Consider the following code:

``` rust
if mmap_size as usize < std::mem::size_of::<run>() {
    ...
}
```

This code has a precedence issue: `as` binds looser than `<`, so I think this attempts to parse what comes after the `as` as a type, `usize<std::mem::size_of ...`, parsing the `<` as a generic.  That results in the following error message:

``` text
error: expected identifier, found `<`
   --> src/main.rs:202:48
    |
202 |     if mmap_size as usize < std::mem::size_of::<run>() {
    |                                                ^
```

This leads the user down the wrong path.  Ideally, rustc should provide an error that hints at the precedence issue instead.

**Comment by Cobrand** [comment]
Ready-to-paste example : 

```
fn main() {
    let i : i64 = 5 ;
    if i as usize < std::mem::size_of::<i64> {
        println!("Case A");
    }
    if i as usize < 5 {
        println!("Case B");
    }
}
```

**Comment by Cobrand** [comment]
I just discovered that "usize", "i64", and stuff like this can actually be a variable name. This might why the error is so confusing.

**Comment by mcarton** [comment]
See also https://github.com/rust-lang/rust/pull/36206.

**Comment by Mark-Simulacrum** [comment]
We spit out the garbage below today. Presumably, this means that usize<...> is what's being parsed.

```
error: unexpected token: `::`
 --> test.rs:3:38
  |
3 |     if i as usize < std::mem::size_of::<i64>() {
  |                                      ^^
  |
  = help: use `<...>` instead of `::<...>` if you meant to specify type arguments

error: expected one of `!`, `+`, `,`, `::`, or `>`, found `(`
 --> test.rs:3:45
  |
3 |     if i as usize < std::mem::size_of::<i64>() {
  |                                             ^ expected one of `!`, `+`, `,`, `::`, or `>` here

error: aborting due to 2 previous errors
```

**Comment by Mark-Simulacrum** [comment]
Closing as fixed since we suggest what's happening today.

```
error: unexpected token: `::`
 --> test.rs:3:38
  |
3 |     if i as usize < std::mem::size_of::<i64>() {
  |                                      ^^
  |
  = help: use `<...>` instead of `::<...>` if you meant to specify type arguments

error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
 --> test.rs:3:45
  |
3 |     if i as usize < std::mem::size_of::<i64>() {
  |                   -                         ^ interpreted as generic argument
  |                   |
  |                   not interpreted as comparison
  |
help: if you want to compare the casted value then write:
  |     if (i as usize) < std::mem::size_of::<i64>() {

error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
 --> test.rs:6:21
  |
6 |     if i as usize < 5 {
  |                   - ^ interpreted as generic argument
  |                   |
  |                   not interpreted as comparison
  |
help: if you want to compare the casted value then write:
  |     if (i as usize) < 5 {

error: aborting due to previous error(s)
```
